File: frontend/types/comparison.ts
import type { ModelType, TransformationOperation, VariableSelectionRequest } from "./modeling";

export type ComparisonStatus = "completed" | "failed" | "unavailable";

export interface ModelFitSummary {
  r_squared: number | null;
  adj_r_squared: number | null;
  within_r_squared: number | null;
  between_r_squared: number | null;
  overall_r_squared: number | null;
  aic: number | null;
  bic: number | null;
  f_statistic: number | null;
  n_obs: number | null;
  n_entities: number | null;
  n_time_periods: number | null;
}

export interface DiagnosticSummary {
  max_vif: number | null;
  heteroskedasticity_detected: boolean | null;
  hausman_rejects_re: boolean | null;
  hausman_p_value: number | null;
  durbin_watson: number | null;
}

export interface CoefficientStabilityEntry {
  model_type: ModelType;
  model_label: string;
  coefficient: number | null;
  std_error: number | null;
  p_value: number | null;
  ci_lower: number | null;
  ci_upper: number | null;
  significance: string;
  direction: "positive" | "negative" | "zero" | "unavailable";
}

export interface ModelComparisonEntry {
  model_type: ModelType;
  model_label: string;
  status: ComparisonStatus;
  reason: string | null;
  formula: string | null;
  fit_metrics: ModelFitSummary | null;
  diagnostic_summary: DiagnosticSummary | null;
  standard_error_type: string | null;
  entity_effects: boolean | null;
  time_effects: boolean | null;
  warnings: string[];
}

export interface ModelScoreComponent {
  criterion: string;
  score: number;
  weight: number;
  explanation: string;
}

export interface ModelSelectionRecommendation {
  recommended_model: ModelType;
  confidence: "high" | "medium" | "low";
  score: number;
  reasons: string[];
  warnings: string[];
  score_breakdown: ModelScoreComponent[];
  alternative_models: { model_type: ModelType; model_label: string; score: string; reason: string }[];
}

export interface ComparisonResult {
  comparison_id: string;
  dataset_id: string;
  dataset_filename: string;
  created_at: string;
  variable_selection: VariableSelectionRequest;
  transformation_summary: string;
  models: ModelComparisonEntry[];
  coefficient_stability: CoefficientStabilityEntry[];
  recommendation: ModelSelectionRecommendation;
  disclaimer: string;
}

export interface ComparisonRequest {
  dataset_id: string;
  variable_selection: VariableSelectionRequest;
  transformations: TransformationOperation[];
  candidate_models: ModelType[];
  cluster_standard_errors_by_entity: boolean;
}

export interface ReportArtifact {
  report_id: string;
  source_type: "analysis" | "comparison";
  source_id: string;
  title: string;
  research_question: string | null;
  created_at: string;
  significance_level: number;
  sections_included: string[];
  markdown_content: string;
  html_content: string;
  writing_rules_version: string;
  disclaimer: string;
}

export interface ReportGenerationRequest {
  source_type: "analysis" | "comparison";
  source_id: string;
  title?: string;
  research_question?: string;
  significance_level?: number;
  include_appendix?: boolean;
}

File: backend/app/schemas/comparison.py
"""Pydantic schemas for multi-model comparison."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.modeling import ModelType, TransformationOperation, VariableSelectionRequest

ComparisonStatus = Literal["completed", "failed", "unavailable"]


class ModelFitSummary(BaseModel):
    r_squared: float | None = None
    adj_r_squared: float | None = None
    within_r_squared: float | None = None
    between_r_squared: float | None = None
    overall_r_squared: float | None = None
    aic: float | None = None
    bic: float | None = None
    f_statistic: float | None = None
    n_obs: int | None = None
    n_entities: int | None = None
    n_time_periods: int | None = None


class DiagnosticSummary(BaseModel):
    max_vif: float | None = None
    heteroskedasticity_detected: bool | None = None
    hausman_rejects_re: bool | None = None
    hausman_p_value: float | None = None
    durbin_watson: float | None = None


class CoefficientStabilityEntry(BaseModel):
    """How the primary IV coefficient behaves across completed models."""

    model_type: ModelType
    model_label: str
    coefficient: float | None
    std_error: float | None
    p_value: float | None
    ci_lower: float | None
    ci_upper: float | None
    significance: str
    direction: str  # "positive", "negative", "zero", "unavailable"


class ModelComparisonEntry(BaseModel):
    model_type: ModelType
    model_label: str
    status: ComparisonStatus
    reason: str | None = None
    formula: str | None = None
    fit_metrics: ModelFitSummary | None = None
    diagnostic_summary: DiagnosticSummary | None = None
    standard_error_type: str | None = None
    entity_effects: bool | None = None
    time_effects: bool | None = None
    warnings: list[str] = Field(default_factory=list)


class ModelScoreComponent(BaseModel):
    criterion: str
    score: int
    weight: float
    explanation: str


class ModelSelectionRecommendation(BaseModel):
    recommended_model: ModelType
    confidence: Literal["high", "medium", "low"]
    score: int
    reasons: list[str]
    warnings: list[str]
    score_breakdown: list[ModelScoreComponent]
    alternative_models: list[dict[str, str]] = Field(default_factory=list)


class ComparisonRequest(BaseModel):
    dataset_id: str
    variable_selection: VariableSelectionRequest
    transformations: list[TransformationOperation] = Field(default_factory=list)
    candidate_models: list[ModelType]
    cluster_standard_errors_by_entity: bool = False


class ComparisonResult(BaseModel):
    comparison_id: str
    dataset_id: str
    dataset_filename: str
    created_at: datetime
    variable_selection: VariableSelectionRequest
    transformation_summary: str
    models: list[ModelComparisonEntry]
    coefficient_stability: list[CoefficientStabilityEntry]
    recommendation: ModelSelectionRecommendation
    disclaimer: str


class ComparisonExportArtifact(BaseModel):
    comparison_id: str
    dataset_id: str
    dataset_filename: str
    created_at: datetime
    variable_selection: dict[str, Any]
    transformations: list[dict[str, Any]]
    candidate_models: list[str]
    model_results: list[dict[str, Any]]
    recommendation: dict[str, Any]
    software_versions: dict[str, str]
    disclaimer: str
